audio-echo/app/src/main/java/com/google/sample/echo/MainActivity.java

```diff
@@ -85,32 +95,13 @@ public boolean onOptionsItemSelected(MenuItem item) {
     }
 
     public void startEcho(View view) {
-        if(!supportRecording || isPlaying) {
-            return;
-        }
-        if(!createSLBufferQueueAudioPlayer()) {
-            status_view.setText("Failed to create Audio Player");
-            return;
-        }
-        if(!createAudioRecorder()) {
-            deleteSLBufferQueueAudioPlayer();
-            status_view.setText("Failed to create Audio Recorder");
-            return;
-        }
-        startPlay();   //this must include startRecording()
-        isPlaying = true;
-        status_view.setText("Engine Echoing ....");
+        status_view.setText("StartCapture Button Clicked\n");
+       startEchoProcessing();
     }
```

Comment:
would it be possible to do the permission at the place calling startEcho(): the picture would be --
   1) user wants to echo, click a button  ( on Java side )
   2) app requests permission ( on Java side )
   3) user grants permission ( on Java side )
   4) app calls startEcho()  to trigger jni to begin (on Java side ); after triggering, the work is done on JNI side for audio

if user does not grant the access, do not bother to call into jni at all, so it stays inside Java side to get work done

vs. right now:
    1) startEchoProcess   ( Java ==> JNI )
    2) calling back to Java for permission ( JNI ==> Java )
    3) Ask for permission ( JAVA )
    4) granted permission ( JAVA )
    5) call back to JNI ( JAVA ==> JNI )
    6) start echoing; from here it is on JNI side